### PR TITLE
Tweak: component invalid address check

### PR DESCRIPTION
**Pull request by shhdgit**
We need to check if the component address is valid to avoid potential security issues.

### CODE REVIEW

**Comment by shhdgit** [comment]
/cc @baurine @breeswish 

PTAL, thanks~

**Review by baurine — APPROVED** [review]
lgtm

File: pkg/tiflash/client.go [hunk 1]
```diff
@@ -79,6 +94,40 @@ func (c *Client) SendGetRequest(host string, statusPort int, relativeURI string)
...
+	err := c.checkValidAddress(fmt.Sprintf("%s:%d", host, statusPort))
+	if err != nil {
+		return nil, err
+	}
+
 	uri := fmt.Sprintf("%s://%s:%d%s", c.httpScheme, host, statusPort, relativeURI)
 	return c.httpClient.WithTimeout(c.timeout).SendRequest(c.lifecycleCtx, uri, http.MethodPost, body, ErrFlashClientRequestFailed, distro.Data("tiflash"))
 }
+
+func (c *Client) getMemebers() ([]topology.StoreInfo, error) {
```
Comment (breeswish) [inline pkg/tiflash/client.go:106]:
```suggestion
func (c *Client) getMembers() ([]topology.StoreInfo, error) {
```

File: pkg/tiflash/client.go [hunk 2]
```diff
@@ -79,6 +94,40 @@ func (c *Client) SendGetRequest(host string, statusPort int, relativeURI string)
...
+	cached, _ := c.cache.Get("tiflash_members")
```
Comment (breeswish) [inline pkg/tiflash/client.go:107]: Seems that each get resets the TTL according to the [doc](https://github.com/ReneKroon/ttlcache/). Maybe this is not a desired behaviour.

File: pkg/tikv/client.go [hunk 3]
```diff
@@ -79,6 +94,40 @@ func (c *Client) SendGetRequest(host string, statusPort int, relativeURI string)
...
 	return c.httpClient.WithTimeout(c.timeout).SendRequest(c.lifecycleCtx, uri, http.MethodPost, body, ErrTiKVClientRequestFailed, distro.Data("tikv"))
 }
+
+func (c *Client) getMemebers() ([]topology.StoreInfo, error) {
+	cached, _ := c.cache.Get("tikv_members")
+	if cached != nil {
+		return cached.([]topology.StoreInfo), nil
+	}
+
+	mem, _, err := topology.FetchStoreTopology(c.pdClient)
```
Comment (breeswish) [inline pkg/tikv/client.go:112]: Only TiKV stores should be allowed.

File: pkg/tiflash/client.go [hunk 4]
```diff
@@ -79,6 +94,40 @@ func (c *Client) SendGetRequest(host string, statusPort int, relativeURI string)
...
+	if cached != nil {
+		return cached.([]topology.StoreInfo), nil
+	}
+
+	_, mem, err := topology.FetchStoreTopology(c.pdClient)
```
Comment (breeswish) [inline pkg/tiflash/client.go:112]: Only TiFlash stores should be allowed.

File: pkg/pd/client.go [hunk 5]
```diff
@@ -38,18 +44,22 @@ const (
...
 	timeout      time.Duration
+	cache        *ttlcache.Cache
 }
 
 func NewPDClient(lc fx.Lifecycle, httpClient *httpc.Client, config *config.Config) *Client {
 	client := &Client{
 		httpClient:   httpClient,
 		httpScheme:   config.GetClusterHTTPScheme(),
 		baseURL:      config.PDEndPoint,
+		specificURL:  "",
```
Comment (breeswish) [inline pkg/pd/client.go:59]: What is this hack?

File: pkg/tiflash/client.go (hunk 2 above)
Comment (shhdgit) [reply pkg/tiflash/client.go:107]: That's right. I should set `SkipTTLExtensionOnHit` to true.

File: pkg/pd/client.go (hunk 5 above)
Comment (shhdgit) [reply pkg/pd/client.go]: Updated. This seems like an unnecessary hack.

File: pkg/tiflash/client.go (hunk 4 above)
Comment (shhdgit) [reply pkg/tiflash/client.go]: What does this mean? And I updated the variable name. PTAL~

File: pkg/tidb/client.go [hunk 6]
```diff
@@ -66,22 +67,28 @@ func NewTiDBClient(lc fx.Lifecycle, config *config.Config, etcdClient *clientv3.
...
 		sqlAPITLSKey:             sqlAPITLSKey,
 		sqlAPIAddress:            "",
-		cache:                    ttlcache.NewCache(),
+		cache:                    cache,
 	}
 
 	lc.Append(fx.Hook{
 		OnStart: func(ctx context.Context) error {
 			client.lifecycleCtx = ctx
+			statusProxyIP, statusProxyPort, _ := host.ParseHostAndPortFromAddressURL("http://" + client.forwarder.statusProxy.listener.Addr().String())
```
Comment (breeswish) [inline pkg/tidb/client.go:89]: Will this cause problems under TLS?

**Review by breeswish — APPROVED** [review]

File: pkg/tidb/client.go (hunk 6 above)
Comment (shhdgit) [reply pkg/tidb/client.go]: Nope, but I found it unnecessary to parse the `net.Addr`.

**Pull request merged**